api/src/main/java/org/eclipse/microprofile/metrics/annotation/ConcurrentGauge.java

```diff
@@ -0,0 +1,138 @@
+ *  Copyright (c) 2018 Contributors to the Eclipse Foundation
+ *
+ *  See the NOTICES file(s) distributed with this work for additional
+ *  information regarding copyright ownership.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  You may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ * ********************************************************************
+ *
+ */
+package org.eclipse.microprofile.metrics.annotation;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Inherited;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import javax.enterprise.util.Nonbinding;
+import javax.interceptor.InterceptorBinding;
+import org.eclipse.microprofile.metrics.MetricUnits;
+
+/**
+ * An annotation for marking a method, constructor, or class as concurrent gauged.
+ * The metric will be registered in the application MetricRegistry.
+ *
+ * A concurrent gauge has the following semantics:
+ * <ul>
+ *   <li>Upon entering the marked item, the value is increased.</li>
+ *   <li>Upon exiting the marked item, the value is decreased.</li>
+ * </ul>
+ *
+ * <p>
+ * Given a method annotated with {@literal @}ConcurrentGauge like this:
+ * </p>
+ * <pre><code>
+ *     {@literal @}ConcurrentGauge(name = "fancyName")
+ *     public String fancyName(String name) {
+ *         return "Sir Captain " + name;
+ *     }
+ * </code></pre>
+ * A concurrent gauge with the fully qualified class name + {@code fancyName} will be created and each time the
+ * {@code #fancyName(String)} method is invoked, the counter will be marked.
+ * Similarly, the same applies for a constructor annotated with counted.
+ *
+ * <p>
+ * Given a class annotated with {@literal @}ConcurrentGauge like this:
+ * </p>
+ * <pre><code>
+ *     {@literal @}ConcurrentGauge
+ *     public class CounterBean {
+ *         public void countMethod1() {}
+ *         public void countMethod2() {}
+ *     }
+ * </code></pre>
+ * A counter for the defining class will be created for each of the constructors/methods.
+ * Each time the constructor/method is invoked, the respective counter will be marked.
+ *
+ * @since 2.0
+ */
+@Inherited
+@Documented
+@InterceptorBinding
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ ElementType.TYPE, ElementType.CONSTRUCTOR, ElementType.METHOD, ElementType.ANNOTATION_TYPE })
```

Comment:
I am not sure why ElementType.ANNOTATION_TYPE should be included. I cannot find a use case for such an annotation class to have this binding applied.

Reply:
TBH I have no idea :-)  I copied it from `Counted`, but I guess it does not make sense there either. But then what do I know :)

Reply:
Having annotations on other annotations makes sense for CDI stereotypes. However, I'm not sure the stereotypes are applicable to interceptor bindings, maybe @antoinesd knows?

Reply:
Apparently according to the Java EE tutorial, it makes sense to add interceptors to stereotype annotations: https://docs.oracle.com/javaee/6/tutorial/doc/gkhqc.html
So having `ElementType.ANNOTATION_TYPE` here really makes sense.

Reply:
Not really. You see ApplicationScoped has the target of @Target({ TYPE, METHOD, FIELD }). It can be part of stereotype without any problem. Type means class, which includes annotation type. Therefore, I think we should not include the ANNOTATION_TYPE.

Reply:
You're right, @Emily-Jiang, thanks for correcting me. I studied a bit more and ANNOTATION_TYPE seems to be a subset of TYPE. It makes sense only if the usage should be restricted to annotations (for meta annotations) and not other types. Having TYPE already allows using an annotation with other annotations and it's possible to add them to a stereotype.

We can remove ANNOTATION_TYPE as @Emily-Jiang suggested.